File: chroma-league/src/main/java/net/booone/chromaleague/hud/parts/health/GainedHealthAnimation.java
package net.booone.chromaleague.hud.parts.health;

import net.booone.chromaleague.hud.colors.CLColor;
import net.booone.razersdk.animation.AnimatedFrame;
import net.booone.razersdk.animation.SimpleFrame;
import net.booone.razersdk.color.StaticColor;
import net.booone.razersdk.color.TransitionColor;
import net.booone.razersdk.sdk.RzKey;

import java.util.List;
import java.util.stream.IntStream;

import static net.booone.chromaleague.hud.parts.health.HealthBar.getHpBarPart;

public class GainedHealthAnimation extends AnimatedFrame {
    private static final int GAINED_HEALTH_CHANGE_STEPS = 5;

    public GainedHealthAnimation(double previousHp, double currentHp) {
        createGainedHealthAnimation(previousHp, currentHp);
    }

    private void createGainedHealthAnimation(double previousHp, double currentHp) {
        final List<RzKey> rzKeys = computeGainedHealth(previousHp, currentHp);
        if (!rzKeys.isEmpty()) {
            final TransitionColor transitionColor = new TransitionColor(StaticColor.WHITE, CLColor.HEALTH, GAINED_HEALTH_CHANGE_STEPS);
            IntStream.range(0, GAINED_HEALTH_CHANGE_STEPS).forEach(i -> addAnimationFrame(new SimpleFrame(rzKeys, transitionColor.getColor())));
        }
    }

    private static List<RzKey> computeGainedHealth(double previousHp, double currentHp) {
        return getHpBarPart(previousHp, currentHp);
    }
}

File: src/hud/parts/health/GainedHealthAnimation.ts
import {AnimatedFrame} from '../../../razer-sdk/animation/AnimatedFrame.js';
import {SimpleFrame} from '../../../razer-sdk/animation/SimpleFrame.js';
import {StaticColor} from '../../../razer-sdk/color/StaticColor.js';
import {TransitionColor} from '../../../razer-sdk/color/TransitionColor.js';
import {RzKey} from '../../../razer-sdk/sdk/RzKey.js';
import {CLColor} from '../../colors/CLColor.js';
import {getHpBarPart} from './HealthBar.js';

const GAINED_HEALTH_CHANGE_STEPS = 5;

export class GainedHealthAnimation extends AnimatedFrame {
    constructor(previousHp: number, currentHp: number) {
        super();
        this.createGainedHealthAnimation(previousHp, currentHp);
    }

    private createGainedHealthAnimation(previousHp: number, currentHp: number): void {
        const rzKeys = computeGainedHealth(previousHp, currentHp);
        if (rzKeys.length > 0) {
            const transitionColor = new TransitionColor(StaticColor.WHITE, CLColor.HEALTH, GAINED_HEALTH_CHANGE_STEPS);
            for (let i = 0; i < GAINED_HEALTH_CHANGE_STEPS; i++) {
                this.addAnimationFrame(new SimpleFrame(rzKeys, transitionColor.getColor()));
            }
        }
    }
}

function computeGainedHealth(previousHp: number, currentHp: number): RzKey[] {
    return getHpBarPart(previousHp, currentHp);
}
